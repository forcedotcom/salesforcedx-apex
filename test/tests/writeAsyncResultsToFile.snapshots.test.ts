/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { join } from 'path';
import { mkdir, readFile, rm } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { tmpdir } from 'os';
import matchSnapshot from 'mocha-snap';
import {
  writeAsyncResultsToFile,
  ApexTestResultOutcome,
  TestResult
} from '../../src';

describe('writeAsyncResultsToFile - Snapshot Tests', () => {
  let tempDir: string;

  beforeEach(async function () {
    tempDir = join(
      tmpdir(),
      `apex-async-snapshot-test-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`
    );
    await mkdir(tempDir, { recursive: true });
  });

  afterEach(async () => {
    if (existsSync(tempDir)) {
      await rm(tempDir, { recursive: true, force: true });
    }
  });

  const mockSnapshotResults: TestResult = {
    summary: {
      outcome: 'Completed',
      testsRan: 4,
      passing: 3,
      failing: 1,
      skipped: 0,
      passRate: '75%',
      failRate: '25%',
      skipRate: '0%',
      testStartTime: '2023-01-01T14:30:00.000Z',
      testExecutionTimeInMs: 2000,
      testTotalTimeInMs: 3000,
      commandTimeInMs: 4000,
      hostname: 'async-snapshot-host',
      orgId: '00D000000000000EAA',
      username: 'async-snapshot@example.com',
      testRunId: 'async-snapshot-run-789',
      userId: '005000000000000AAA'
    },
    tests: [
      {
        id: '01p000000000001AAA',
        queueItemId: '709000000000001AAA',
        stackTrace: null as string | null,
        message: null as string | null,
        asyncApexJobId: '707000000000001AAA',
        methodName: 'testAsyncSnapshotMethod1',
        outcome: ApexTestResultOutcome.Pass,
        apexLogId: null as string | null,
        apexClass: {
          id: '01p000000000001AAA',
          name: 'AsyncSnapshotTestClass1',
          namespacePrefix: null as string | null,
          fullName: 'AsyncSnapshotTestClass1'
        },
        runTime: 800,
        testTimestamp: '2023-01-01T14:30:00.500Z',
        fullName: 'AsyncSnapshotTestClass1.testAsyncSnapshotMethod1'
      },
      {
        id: '01p000000000002AAA',
        queueItemId: '709000000000002AAA',
        stackTrace:
          'System.AsyncException: Async operation failed\\nLine 15: someMethod()\\nLine 23: anotherMethod()',
        message: 'Async test failed with complex error details',
        asyncApexJobId: '707000000000001AAA',
        methodName: 'testAsyncSnapshotMethod2',
        outcome: ApexTestResultOutcome.Fail,
        apexLogId: '07L000000000001AAA',
        apexClass: {
          id: '01p000000000002AAA',
          name: 'AsyncSnapshotTestClass2',
          namespacePrefix: 'AsyncNamespace',
          fullName: 'AsyncNamespace.AsyncSnapshotTestClass2'
        },
        runTime: 600,
        testTimestamp: '2023-01-01T14:30:01.100Z',
        fullName:
          'AsyncNamespace.AsyncSnapshotTestClass2.testAsyncSnapshotMethod2'
      },
      {
        id: '01p000000000003AAA',
        queueItemId: '709000000000003AAA',
        stackTrace: null as string | null,
        message: null as string | null,
        asyncApexJobId: '707000000000001AAA',
        methodName: 'testAsyncSnapshotMethod3',
        outcome: ApexTestResultOutcome.Pass,
        apexLogId: null as string | null,
        apexClass: {
          id: '01p000000000003AAA',
          name: 'AsyncSnapshotTestClass3',
          namespacePrefix: null as string | null,
          fullName: 'AsyncSnapshotTestClass3'
        },
        runTime: 400,
        testTimestamp: '2023-01-01T14:30:01.500Z',
        fullName: 'AsyncSnapshotTestClass3.testAsyncSnapshotMethod3'
      },
      {
        id: '01p000000000004AAA',
        queueItemId: '709000000000004AAA',
        stackTrace: null as string | null,
        message: null as string | null,
        asyncApexJobId: '707000000000001AAA',
        methodName: 'testAsyncSnapshotMethod4',
        outcome: ApexTestResultOutcome.Pass,
        apexLogId: null as string | null,
        apexClass: {
          id: '01p000000000004AAA',
          name: 'AsyncSnapshotTestClass4',
          namespacePrefix: 'AnotherNamespace',
          fullName: 'AnotherNamespace.AsyncSnapshotTestClass4'
        },
        runTime: 200,
        testTimestamp: '2023-01-01T14:30:01.700Z',
        fullName:
          'AnotherNamespace.AsyncSnapshotTestClass4.testAsyncSnapshotMethod4'
      }
    ]
  };

  it('should produce consistent async results JSON output', async function () {
    const runId = 'snapshot-async-test-123';

    await writeAsyncResultsToFile(mockSnapshotResults, runId);

    const expectedPath = join(tmpdir(), runId, 'rawResults.json');
    const content = await readFile(expectedPath, 'utf8');
    const parsedContent = JSON.parse(content);

    // Snapshot the complete async results structure
    matchSnapshot(parsedContent);

    // Cleanup
    await rm(join(tmpdir(), runId), { recursive: true, force: true });
  });

  it('should produce consistent output with null values', async function () {
    const resultsWithNulls: TestResult = {
      summary: {
        ...mockSnapshotResults.summary,
        testRunId: 'null-values-snapshot-test'
      },
      tests: [
        {
          id: '01p000000000001AAA',
          queueItemId: '709000000000001AAA',
          stackTrace: null as string | null,
          message: null as string | null,
          asyncApexJobId: '707000000000001AAA',
          methodName: 'testWithNullValues',
          outcome: ApexTestResultOutcome.Pass,
          apexLogId: null as string | null,
          apexClass: {
            id: '01p000000000001AAA',
            name: 'NullValuesTestClass',
            namespacePrefix: null as string | null,
            fullName: 'NullValuesTestClass'
          },
          runTime: 250,
          testTimestamp: '2023-01-01T14:30:00.000Z',
          fullName: 'NullValuesTestClass.testWithNullValues'
        }
      ]
    };

    const runId = 'null-values-snapshot';

    await writeAsyncResultsToFile(resultsWithNulls, runId);

    const expectedPath = join(tmpdir(), runId, 'rawResults.json');
    const content = await readFile(expectedPath, 'utf8');
    const parsedContent = JSON.parse(content);

    // Snapshot to ensure null values are handled consistently
    matchSnapshot(parsedContent);

    // Cleanup
    await rm(join(tmpdir(), runId), { recursive: true, force: true });
  });

  it('should produce consistent output with complex nested data', async function () {
    const complexResults: TestResult = {
      ...mockSnapshotResults,
      tests: [
        {
          ...mockSnapshotResults.tests[0],
          perClassCoverage: [] // Test empty arrays
        },
        {
          ...mockSnapshotResults.tests[1],
          stackTrace:
            'Multi-line\\nstack\\ntrace\\nwith\\nspecial\\ncharacters: "quotes", \\backslashes, and [brackets]'
        }
      ]
    };

    const runId = 'complex-data-snapshot';

    await writeAsyncResultsToFile(complexResults, runId);

    const expectedPath = join(tmpdir(), runId, 'rawResults.json');
    const content = await readFile(expectedPath, 'utf8');
    const parsedContent = JSON.parse(content);

    // Snapshot to ensure complex data structures are handled consistently
    matchSnapshot(parsedContent);

    // Cleanup
    await rm(join(tmpdir(), runId), { recursive: true, force: true });
  });

  it('should produce consistent output with empty test results', async function () {
    const emptyResults: TestResult = {
      summary: {
        outcome: 'Completed',
        testsRan: 0,
        passing: 0,
        failing: 0,
        skipped: 0,
        passRate: '0%',
        failRate: '0%',
        skipRate: '0%',
        testStartTime: '2023-01-01T14:30:00.000Z',
        testExecutionTimeInMs: 0,
        testTotalTimeInMs: 0,
        commandTimeInMs: 100,
        hostname: 'empty-test-host',
        orgId: '00D000000000000EAA',
        username: 'empty@example.com',
        testRunId: 'empty-snapshot-test',
        userId: '005000000000000AAA'
      },
      tests: []
    };

    const runId = 'empty-results-snapshot';

    await writeAsyncResultsToFile(emptyResults, runId);

    const expectedPath = join(tmpdir(), runId, 'rawResults.json');
    const content = await readFile(expectedPath, 'utf8');
    const parsedContent = JSON.parse(content);

    // Snapshot to ensure empty results are handled consistently
    matchSnapshot(parsedContent);

    // Cleanup
    await rm(join(tmpdir(), runId), { recursive: true, force: true });
  });
});
