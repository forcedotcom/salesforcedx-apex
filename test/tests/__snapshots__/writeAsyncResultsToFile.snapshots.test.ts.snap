// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeAsyncResultsToFile - Snapshot Tests should produce consistent async results JSON output 1`] = `
Object {
  "summary": Object {
    "commandTimeInMs": 4000,
    "failRate": "25%",
    "failing": 1,
    "hostname": "async-snapshot-host",
    "orgId": "00D000000000000EAA",
    "outcome": "Completed",
    "passRate": "75%",
    "passing": 3,
    "skipRate": "0%",
    "skipped": 0,
    "testExecutionTimeInMs": 2000,
    "testRunId": "async-snapshot-run-789",
    "testStartTime": "2023-01-01T14:30:00.000Z",
    "testTotalTimeInMs": 3000,
    "testsRan": 4,
    "userId": "005000000000000AAA",
    "username": "async-snapshot@example.com",
  },
  "tests": Array [
    Object {
      "apexClass": Object {
        "fullName": "AsyncSnapshotTestClass1",
        "id": "01p000000000001AAA",
        "name": "AsyncSnapshotTestClass1",
        "namespacePrefix": null,
      },
      "apexLogId": null,
      "asyncApexJobId": "707000000000001AAA",
      "fullName": "AsyncSnapshotTestClass1.testAsyncSnapshotMethod1",
      "id": "01p000000000001AAA",
      "message": null,
      "methodName": "testAsyncSnapshotMethod1",
      "outcome": "Pass",
      "queueItemId": "709000000000001AAA",
      "runTime": 800,
      "stackTrace": null,
      "testTimestamp": "2023-01-01T14:30:00.500Z",
    },
    Object {
      "apexClass": Object {
        "fullName": "AsyncNamespace.AsyncSnapshotTestClass2",
        "id": "01p000000000002AAA",
        "name": "AsyncSnapshotTestClass2",
        "namespacePrefix": "AsyncNamespace",
      },
      "apexLogId": "07L000000000001AAA",
      "asyncApexJobId": "707000000000001AAA",
      "fullName": "AsyncNamespace.AsyncSnapshotTestClass2.testAsyncSnapshotMethod2",
      "id": "01p000000000002AAA",
      "message": "Async test failed with complex error details",
      "methodName": "testAsyncSnapshotMethod2",
      "outcome": "Fail",
      "queueItemId": "709000000000002AAA",
      "runTime": 600,
      "stackTrace": "System.AsyncException: Async operation failed\\\\nLine 15: someMethod()\\\\nLine 23: anotherMethod()",
      "testTimestamp": "2023-01-01T14:30:01.100Z",
    },
    Object {
      "apexClass": Object {
        "fullName": "AsyncSnapshotTestClass3",
        "id": "01p000000000003AAA",
        "name": "AsyncSnapshotTestClass3",
        "namespacePrefix": null,
      },
      "apexLogId": null,
      "asyncApexJobId": "707000000000001AAA",
      "fullName": "AsyncSnapshotTestClass3.testAsyncSnapshotMethod3",
      "id": "01p000000000003AAA",
      "message": null,
      "methodName": "testAsyncSnapshotMethod3",
      "outcome": "Pass",
      "queueItemId": "709000000000003AAA",
      "runTime": 400,
      "stackTrace": null,
      "testTimestamp": "2023-01-01T14:30:01.500Z",
    },
    Object {
      "apexClass": Object {
        "fullName": "AnotherNamespace.AsyncSnapshotTestClass4",
        "id": "01p000000000004AAA",
        "name": "AsyncSnapshotTestClass4",
        "namespacePrefix": "AnotherNamespace",
      },
      "apexLogId": null,
      "asyncApexJobId": "707000000000001AAA",
      "fullName": "AnotherNamespace.AsyncSnapshotTestClass4.testAsyncSnapshotMethod4",
      "id": "01p000000000004AAA",
      "message": null,
      "methodName": "testAsyncSnapshotMethod4",
      "outcome": "Pass",
      "queueItemId": "709000000000004AAA",
      "runTime": 200,
      "stackTrace": null,
      "testTimestamp": "2023-01-01T14:30:01.700Z",
    },
  ],
}
`;

exports[`writeAsyncResultsToFile - Snapshot Tests should produce consistent output with complex nested data 1`] = `
Object {
  "summary": Object {
    "commandTimeInMs": 4000,
    "failRate": "25%",
    "failing": 1,
    "hostname": "async-snapshot-host",
    "orgId": "00D000000000000EAA",
    "outcome": "Completed",
    "passRate": "75%",
    "passing": 3,
    "skipRate": "0%",
    "skipped": 0,
    "testExecutionTimeInMs": 2000,
    "testRunId": "async-snapshot-run-789",
    "testStartTime": "2023-01-01T14:30:00.000Z",
    "testTotalTimeInMs": 3000,
    "testsRan": 4,
    "userId": "005000000000000AAA",
    "username": "async-snapshot@example.com",
  },
  "tests": Array [
    Object {
      "apexClass": Object {
        "fullName": "AsyncSnapshotTestClass1",
        "id": "01p000000000001AAA",
        "name": "AsyncSnapshotTestClass1",
        "namespacePrefix": null,
      },
      "apexLogId": null,
      "asyncApexJobId": "707000000000001AAA",
      "fullName": "AsyncSnapshotTestClass1.testAsyncSnapshotMethod1",
      "id": "01p000000000001AAA",
      "message": null,
      "methodName": "testAsyncSnapshotMethod1",
      "outcome": "Pass",
      "perClassCoverage": Array [],
      "queueItemId": "709000000000001AAA",
      "runTime": 800,
      "stackTrace": null,
      "testTimestamp": "2023-01-01T14:30:00.500Z",
    },
    Object {
      "apexClass": Object {
        "fullName": "AsyncNamespace.AsyncSnapshotTestClass2",
        "id": "01p000000000002AAA",
        "name": "AsyncSnapshotTestClass2",
        "namespacePrefix": "AsyncNamespace",
      },
      "apexLogId": "07L000000000001AAA",
      "asyncApexJobId": "707000000000001AAA",
      "fullName": "AsyncNamespace.AsyncSnapshotTestClass2.testAsyncSnapshotMethod2",
      "id": "01p000000000002AAA",
      "message": "Async test failed with complex error details",
      "methodName": "testAsyncSnapshotMethod2",
      "outcome": "Fail",
      "queueItemId": "709000000000002AAA",
      "runTime": 600,
      "stackTrace": "Multi-line\\\\nstack\\\\ntrace\\\\nwith\\\\nspecial\\\\ncharacters: \\"quotes\\", \\\\backslashes, and [brackets]",
      "testTimestamp": "2023-01-01T14:30:01.100Z",
    },
  ],
}
`;

exports[`writeAsyncResultsToFile - Snapshot Tests should produce consistent output with empty test results 1`] = `
Object {
  "summary": Object {
    "commandTimeInMs": 100,
    "failRate": "0%",
    "failing": 0,
    "hostname": "empty-test-host",
    "orgId": "00D000000000000EAA",
    "outcome": "Completed",
    "passRate": "0%",
    "passing": 0,
    "skipRate": "0%",
    "skipped": 0,
    "testExecutionTimeInMs": 0,
    "testRunId": "empty-snapshot-test",
    "testStartTime": "2023-01-01T14:30:00.000Z",
    "testTotalTimeInMs": 0,
    "testsRan": 0,
    "userId": "005000000000000AAA",
    "username": "empty@example.com",
  },
  "tests": Array [],
}
`;

exports[`writeAsyncResultsToFile - Snapshot Tests should produce consistent output with null values 1`] = `
Object {
  "summary": Object {
    "commandTimeInMs": 4000,
    "failRate": "25%",
    "failing": 1,
    "hostname": "async-snapshot-host",
    "orgId": "00D000000000000EAA",
    "outcome": "Completed",
    "passRate": "75%",
    "passing": 3,
    "skipRate": "0%",
    "skipped": 0,
    "testExecutionTimeInMs": 2000,
    "testRunId": "null-values-snapshot-test",
    "testStartTime": "2023-01-01T14:30:00.000Z",
    "testTotalTimeInMs": 3000,
    "testsRan": 4,
    "userId": "005000000000000AAA",
    "username": "async-snapshot@example.com",
  },
  "tests": Array [
    Object {
      "apexClass": Object {
        "fullName": "NullValuesTestClass",
        "id": "01p000000000001AAA",
        "name": "NullValuesTestClass",
        "namespacePrefix": null,
      },
      "apexLogId": null,
      "asyncApexJobId": "707000000000001AAA",
      "fullName": "NullValuesTestClass.testWithNullValues",
      "id": "01p000000000001AAA",
      "message": null,
      "methodName": "testWithNullValues",
      "outcome": "Pass",
      "queueItemId": "709000000000001AAA",
      "runTime": 250,
      "stackTrace": null,
      "testTimestamp": "2023-01-01T14:30:00.000Z",
    },
  ],
}
`;
